generate-migration-guide/src/main.rs

```diff
@@ -0,0 +1,355 @@
+    path::PathBuf,
+};
+
+/// Generates markdown files used for a bevy releases.
+///
+/// Migration Guide:
+/// * Gets all PRs with the `C-Breaking-Change` label and that were merged by bors.
+/// * For each PR:
+///     * Generate the title with a link to the relevant PR and
+///     * Generate the migration guide section. This parses the markdown and generates valid makrdown that should pass markdownlint rules.
+///
+/// Release notes:
+/// * Gets all PRs merged by bors
+/// * Collect each author of closed PRs (Should this just list all contributors?)
+/// * Sort each PR per area label
+/// * Generate the list of merge PR
+///
+/// Requires a valid GITHUB_TOKEN environment variable, you can use a .env file or use your prefered method of passing env arguments.
+///
+/// Example used to generate for 0.9:
+/// cargo run -- migration-guide --date 2022-07-31 --title "0.8 to 0.9" --weight 5
+/// cargo run -- release-note --date 2022-07-31
+#[derive(ClapParser)]
+#[command(author, version, about)]
+struct Args {
+    #[command(subcommand)]
+    command: Commands,
+}
+
+#[derive(Subcommand)]
+enum Commands {
+    MigrationGuide {
+        /// Date of the release of the previous version. Format: YYYY-MM-DD
+        #[arg(short, long)]
+        date: String,
+
+        /// Title of the frontmatter
+        #[arg(short, long)]
+        title: String,
+
+        /// Weight used for sorting
+        #[arg(short, long)]
+        weight: i32,
+
+        /// Path used to output the generated file. Defaults to ./migration-guide.md
+        #[arg(short, long)]
+        path: Option<std::path::PathBuf>,
+    },
+    ReleaseNote {
+        /// Date of the release of the previous version. Format: YYYY-MM-DD
+        #[arg(short, long)]
+        date: String,
+
+        /// Path used to output the generated file. Defaults to ./release-note.md
+        #[arg(short, long)]
+        path: Option<std::path::PathBuf>,
+    },
+}
+
+fn main() -> anyhow::Result<()> {
+    let _ = dotenv::dotenv();
+    let args = Args::parse();
+    match args.command {
+        Commands::MigrationGuide {
+            date,
+            title,
+            weight,
+            path,
+        } => generate_migration_guide(
+            &title,
+            weight,
+            &date,
+            path.unwrap_or_else(|| PathBuf::from("./migration-guide.md")),
+        )?,
+        Commands::ReleaseNote { date, path } => generate_release_note(
+            &date,
+            path.unwrap_or_else(|| PathBuf::from("./release-note.md")),
+        )?,
+    };
+
+    Ok(())
+}
+
+/// Generates the list of contributors and a list of all closed PRs sorted by area labels
+fn generate_release_note(date: &str, path: PathBuf) -> anyhow::Result<()> {
+    let prs = get_merged_prs(date, None)?;
+    let mut authors = HashSet::new();
+    let mut pr_map = HashMap::new();
+    let mut areas = HashMap::<String, Vec<i32>>::new();
+    for pr in &prs {
+        authors.insert(pr.user.login.clone());
```

Comment:
Another issue is that this excludes "pr collaborators" from the contributors list. We should find some way to include this list of people:
![image](https://user-images.githubusercontent.com/2694663/200431165-64d3cd24-a5d7-48d4-be80-822d631e53d6.png)

Reply:
Yeah, I haven't figured out yet how to get that info. It's not on the issue or PR response from the rest api. I think most people that collaborated on PRs probably also have PRs where they were authors, but there's probably a couple that we missed anyway.

Reply:
Wait, now that I think about it, I can probably just get all the commits on main, that way I don't need to filter anything other than the date since every commit contains the PR description and it's authors.

Reply:
Oooh yeah thats a good call.

Reply:
Hmm so you can list co-authors like this: `git log v0.8.0..origin/main --pretty=format:"%(trailers:key=Co-authored-by)"`, but that will only give you the emails and first / last names (generally), not github accounts (which is what we use).
So far I haven't been able to find any api that provides the co-author github user accounts.

Reply:
I've spent a good portion of time trying to build the `email->username` correlation, but I haven't found anything yet. The "repo contributors" api does have an email field in the schema, but it doesn't actually show up in the response.

There is a "release notes generator" which might have worked :https://docs.github.com/en/enterprise-server@3.5/rest/releases/releases#generate-release-notes-content-for-a-release

However it only includes PRs in the "merged" state, which doesn't work for our needs.

So far the _only_ way I've found to accurately generate this list is by scraping each page on the GitHub "compare branches" view:
https://github.com/bevyengine/bevy/compare/v0.8.0...main
Which I have done for past releases by manually copy/pasting each page's content and then extracting relevant info using regexes.

Reply:
I'm going to do that for this release as I can't afford to spend more time on this.

Reply:
you can use `https://api.github.com/search/users?q=<email>` to find the user

Reply:
I now use the graphql api and this gives us a full list of contributors for each commit